### PR TITLE
optimize profiler

### PR DESCRIPTION
**Pull request by jacquesqiao**

### CODE REVIEW

File: paddle/fluid/platform/device_tracer.cc [hunk 1]
```diff
@@ -332,6 +354,17 @@ class DeviceTracerImpl : public DeviceTracer {
     enabled_ = false;
   }
 
+  inline AllRecord &GetThreadLocalAllRecord() {
+    if (!thread_local_records) {
+      std::lock_guard<std::mutex> guard(trace_mu_);
+      if (!thread_local_records) {
+        thread_local_records = std::make_shared<AllRecord>();
+        g_all_records.push_back(thread_local_records);
```
Comment (panyx0718) [inline paddle/fluid/platform/device_tracer.cc:362]: Is there a doc that shows it's correct? Are thread_local variables auto visible?

File: paddle/fluid/platform/profiler.cc [hunk 2]
```diff
@@ -173,7 +175,6 @@ void PopEvent(const std::string& name, const DeviceContext* dev_ctx) {
 
 RecordEvent::RecordEvent(const std::string& name, const DeviceContext* dev_ctx)
     : is_enabled_(false), start_ns_(PosixInNsec()) {
-  std::lock_guard<std::mutex> l(profiler_mu);
```
Comment (panyx0718) [inline paddle/fluid/platform/profiler.cc]: this seems wrong to me

**Review by panyx0718 — COMMENTED** [review]
what's the improvements?

File: python/paddle/fluid/profiler.py [hunk 3]
```diff
@@ -122,7 +122,7 @@ def reset_profiler():
     core.reset_profiler()
 
 
-def start_profiler(state):
+def start_profiler(state, enable_timeline=True):
```
Comment (Yancey0623) [inline python/paddle/fluid/profiler.py:125]: Can we enable the timeline if call `enable_profiler`?

**Pull request closed without merging**